contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,506 @@
+    address public immutable _eip712StETH;
+    address public immutable _voting;
+    address public immutable _nodeOperatorsRegistry;
+    address public immutable _hashConsensusForAccountingOracle;
+    address public immutable _hashConsensusForValidatorsExitBusOracle;
+    address public immutable _gateSeal;
+    address public immutable _withdrawalQueueImplementation;
+    address public immutable _stakingRouterImplementation;
+    address public immutable _accountingOracleImplementation;
+    address public immutable _validatorsExitBusOracleImplementation;
+    bytes32 public immutable _withdrawalCredentials;
+    uint256 public immutable _nodeOperatorsRegistryStuckPenaltyDelay;
+    address public immutable _dummyImplementation;
+    address public immutable _locatorImplementation;
+
+    //
+    // STRUCTURED STORAGE
+    //
+    bool public isUpgradeStarted;
+    bool public isUpgradeFinished;
+
+    constructor(Config memory _config, ConfigImplementations memory _configImpl) {
+        // TODO mainnet/testnet: update the values
+        _accountingOracleConsensusVersion = 1;
+        _validatorsExitBusOracleConsensusVersion = 1;
+        _nodeOperatorsRegistryStakingModuleType = NOR_STAKING_MODULE_NAME_BYTES;
+
+        // TODO mainnet: hardcode the values
+        _locator = _config.locator;
+        _eip712StETH = _config.eip712StETH;
+        _voting = _config.voting;
+        _nodeOperatorsRegistry = _config.nodeOperatorsRegistry;
+        _hashConsensusForAccountingOracle = _config.hashConsensusForAccountingOracle;
+        _hashConsensusForValidatorsExitBusOracle = _config.hashConsensusForValidatorsExitBusOracle;
+        _gateSeal = _config.gateSeal;
+        _withdrawalCredentials = _config.withdrawalCredentials;
+        _nodeOperatorsRegistryStuckPenaltyDelay = _config.nodeOperatorsRegistryStuckPenaltyDelay;
+        _withdrawalQueueImplementation = _configImpl.withdrawalQueueImplementation;
+        _stakingRouterImplementation = _configImpl.stakingRouterImplementation;
+        _accountingOracleImplementation = _configImpl.accountingOracleImplementation;
+        _validatorsExitBusOracleImplementation = _configImpl.validatorsExitBusOracleImplementation;
+        _dummyImplementation = _configImpl.dummyImplementation;
+        _locatorImplementation = _configImpl.locatorImplementation;
+    }
+
+    function verifyInitialState() external view {
+        _verifyInitialState();
+    }
+
+    /// Need to be called before LidoOracle implementation is upgraded to LegacyOracle
+    function startUpgrade() external {
+        _startUpgrade();
+    }
+
+    function finishUpgrade() external {
+        _finishUpgrade();
+    }
+
+    /// Perform basic checks to revert the entire upgrade if something gone wrong
+    function verifyFinishedUpgrade() external view {
+        _verifyFinishedUpgrade();
+    }
+
+    function _startUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
```

Comment:
maybe place auth section in the external method body?

Reply:
Not sure. Current option seems for me to be less ambiguous: all checks are in a raw. No impression that msg.sender is the only "entrance" check